README.md
![alt tag](https://raw.githubusercontent.com/lateralblast/parsec/master/sparc_t5.png)

PARSEC
======

Parse, Analyse, and Report on Solaris Explorer Configuration/Content

Introduction
------------

A ruby script to parse a Sun Oracle Solaris explorer file and extract information.
There is some reporting capability being built into the script.

Some of the features include:

- Map IO paths and disk
- Determine part numbers where possible
- Report on firmware versions
  - Report if newer version is available
- Security checks for defaults and kernel parameters
- Mask customer information, WWNs, etc
- Output in the following formats:
  - Plain console text which can be piped into other applications
  - Text table console output
  - HTML output
  - Wiki output
- A Sinatra based webserver to serve up configs

Usage
-----

The default output has changed to pipe (white space delimited plain text).

To get the old style terminal tables, use the table output format type, e.g.:

```
$ ./parsec.rb --server=hostname --report=all --format=table
```

To get the html formated output use the html output format type, e.g.:

```
$ ./parsec.rb --server=hostname --report=all --format=html
```

To get pdf output, use the pdf format output type, e.g.:

```
$ ./parsec.rb --server=hostname --report=all --format=pdf
```

To run the Sinatra based webserver:

```
$ ./webserver
```

This will start up a webserver on port 9494

![alt tag](https://raw.githubusercontent.com/lateralblast/parsec/master/webserver_example.png)

Authentication can be enable by setting the enable_auth flag in webserver.rb:

```
enable_auth = true
```

Certificates and keys are stored in the ssl directory:

```
$ ls -l ssl
total 16
-rw-r--r--  1 spindler  staff  1383  9 Aug 23:27 cert.crt
-rw-r--r--  1 spindler  staff   891  9 Aug 23:27 pkey.pem
```

If they don't exist the script will try to create them on startup.

Otherwise use the openssl command:

```
$ cd ssl
$ openssl req -x509 -nodes -days 365 -newkey rsa:1024 -keyout pkey.pem -out cert.crt
```

The server looks for users in a .htpasswd file in the views directory, eg:

```
$ cat ./views/.htpasswd 
spindler:$2y$05$0Ntm14Sl3Ix.qUSd537V8eUuhM6nJiqSM9mB18Pp5bnbg0ZhtxQAq
```

The password algorithm is bcrypt.

Password entries can be created using htpasswd, eg:

```
$ htpasswd -n -b -B spindler test >> ./views/.htpasswd 
```

If you want photos to be displayed, you'll need the <a href="https://github.com/lateralblast/ssp">Sun System Pictures</a> archive.

To enable photos, clone the archive to a location of your choice and symlink it to the photos directory in the root directory of parsec, e.g.:

```
$ cd /export
$ git clone https://github.com/lateralblast/ssp.git
$ cd /export/parsec
$ ln -s /export/ssp/images ./photos
```

Reports
-------

The reporting options include:

- Report on Aggregate information
- Report on Everything
- Report on Coreadm information
- Report on CPU information
- Report on Cron information
- Report on Crypto information
- Report on CUPS information
- Report on Disks
- Report on Dumpadm information
- Report on EEPROM
- Report on Elfsign information
- Report on Explore information
- Report on Firmware information
- Report on FRU information
- Report on Filesystem information
- Report on Firmware information
- Report on FRU information
- Report on Host information
- Report on inetadm information
- Report on inetd information
- Report on inetinit information
- Report on IPMI information
- Report on IO information
- Report on LDom information
- Report on Link information
- Report on Locale information
- Report on Live Upgrade information
- Report on Kernel information
- Report on Keyserv information
- Report on Memory information
- Report on Kernel Module information
- Report on Mount information
- Report on ndd information
- Report on Network information
- Report on NTP information
- Report on OBP information
- Report on OS information
- Report on package information
- Report on PAM information
- Report on Password information
- Report on Patch information
- Report on PCI information
- Report on Power information
- Report on Security information
- Report on Sendmail settings
- Report on Sensor information
- Report on Chassis Serial information
- Report on Component Serial information
- Report on Services information
- Report on SNMP information
- Report on SSH information
- Report on SU information
- Report on Suspend information
- Report on Syslog information
- Report on System information
- Report on Swap information
- Report on TCP information
- Report on Telnet information
- Report on Upgradeable slot information
- Report on UDP information
- Report on VNIC information
- Report on Veritas information
- Report on ZFS information
- Report on Zone information

License
-------

This software is licensed as CC-BA (Creative Commons By Attrbution)

http://creativecommons.org/licenses/by/4.0/legalcode


Documentation
-------------

- [Wiki](https://github.com/lateralblast/parsec/wiki)
- [Features](https://github.com/lateralblast/parsec/wiki/1.-Features)
- [Usage](https://github.com/lateralblast/parsec/wiki/2.-Usage)
- [Requirements](https://github.com/lateralblast/parsec/wiki/6.-Requirements)

Getting Started
---------------

- [Getting started with Parsec](https://github.com/lateralblast/parsec/wiki/3.-Getting-Started)

Examples
--------

General Examples:

- [General Information](4.0.-General-Information)

Hardware Examples:

- [Memory](https://github.com/lateralblast/parsec/wiki/4.1.1.-Memory)
- [IO](https://github.com/lateralblast/parsec/wiki/4.1.2.-IO)
- [Power](https://github.com/lateralblast/parsec/wiki/4.1.3.-Power)
- [Suspend](https://github.com/lateralblast/parsec/wiki/4.1.4.-Suspend)
- [ILOM](https://github.com/lateralblast/parsec/wiki/4.1.5.-ILOM)
- [Disks](https://github.com/lateralblast/parsec/wiki/4.1.6.-Disks)
- [OBP](https://github.com/lateralblast/parsec/wiki/4.1.7.-OBP)
- [FRU](https://github.com/lateralblast/parsec/wiki/4.1.8.-FRU)
- [CPU](https://github.com/lateralblast/parsec/wiki/4.1.9.-CPU)

Software Examples:

- [Explorer](https://github.com/lateralblast/parsec/wiki/4.2.1.-Explorer)
- [Coreadm](https://github.com/lateralblast/parsec/wiki/4.2.2.-Coreadm)
- [Dumpadm](https://github.com/lateralblast/parsec/wiki/4.2.3.-Dumpadm)
- [Packages](https://github.com/lateralblast/parsec/wiki/4.2.4.-Packages)
- [Patches](https://github.com/lateralblast/parsec/wiki/4.2.5.-Patches)

Network Examples:

- [TCP](https://github.com/lateralblast/parsec/wiki/4.3.1.-TCP)
- [UDP](https://github.com/lateralblast/parsec/wiki/4.3.2.-UDP)
- [SSH](https://github.com/lateralblast/parsec/wiki/4.3.3.-SSH)
- [Sendmail](https://github.com/lateralblast/parsec/wiki/4.3.4.-Sendmail)
- [Telnet](https://github.com/lateralblast/parsec/wiki/4.3.5.-Telnet)

Operating System Examples:

- [Kernel](https://github.com/lateralblast/parsec/wiki/4.4.1.-Kernel)
- [Modules](https://github.com/lateralblast/parsec/wiki/4.4.2.-Modules)
- [Locale](https://github.com/lateralblast/parsec/wiki/4.4.3.-Locale)
- [Filesystem](https://github.com/lateralblast/parsec/wiki/4.4.4.-Filesystem)
- [LiveUpgrade](https://github.com/lateralblast/parsec/wiki/4.4.5.-LiveUpgrade)
- [Services](https://github.com/lateralblast/parsec/wiki/4.4.6.-Services)
- [Host](https://github.com/lateralblast/parsec/wiki/4.4.7.-Host)
- [EEPROM](https://github.com/lateralblast/parsec/wiki/4.4.8.-EEPROM)

Virtualisation Examples:

- [LDoms](https://github.com/lateralblast/parsec/wiki/4.5.1.-LDoms)
- [Zones](https://github.com/lateralblast/parsec/wiki/4.5.2.-Zones)

Security Information:

- [System](https://github.com/lateralblast/parsec/wiki/4.6.1.-System)
- [Password](https://github.com/lateralblast/parsec/wiki/4.6.2.-Password)
- [Login](https://github.com/lateralblast/parsec/wiki/4.6.3.Login)
- [Inetd](https://github.com/lateralblast/parsec/wiki/4.6.4.-Inetd)
- [Inetinit](https://github.com/lateralblast/parsec/wiki/4.6.5.-Inetinit)
- [Su](https://github.com/lateralblast/parsec/wiki/4.6.6.-Su)
- [Cron](https://github.com/lateralblast/parsec/wiki/4.6.7.-Cron)
- [Keyserv](https://github.com/lateralblast/parsec/wiki/4.6.8.-Keyserv)

Requirements
------------

Tools:

- Uses pigz if available
- S tar required for Operating Systems other than Solaris
- ImageMagick required for image processing for PDF reports
- libiconv

Required Ruby Gems (required for text reports):

- rubygems
- fileutils
- getopt/std
- pathname
- hex_string
- terminal-table
- pathname
- etc
- unpack
- enumerator
- iconv

Optional Ruby Gems (required for PDF reports)

- prawn
- prawn/table
- fastimage
- nokogiri
- rmagick

Aditional gems required for webserver:

- erb
- sinatra

There is a very basic script to install gems

```
./utils/generic/install_gems.sh
```

A base set of firmware information is provided.
This information is created using [oort](https://github.com/lateralblast/oort).
In order to keep this information up to date, a MOS (My Oracle Support) account is required.

In order to work, parsec requires three directories:

- methods
  - where the parsec modules are located
  - cloning parsec from git will get these
- information
  - where firmware and other information is location
  - parsec comes with some base information which can be updated using the [oort](https://github.com/lateralblast/oort) script
  - see the [Getting started with Parsec](https://github.com/lateralblast/parsec/wiki/3.-Getting-Started) wiki
- firmware (under development)
  - the actual firmware
  - required for some machines to obtain further information
  - can be the repository generated from the [oort](https://github.com/lateralblast/oort) script
  - see the [Getting started with Parsec](https://github.com/lateralblast/parsec/wiki/3.-Getting-Started) wiki


changelog
# Changes
# 0.0.1: Sat 26 Oct 2013 13:34:48 EST
# Initial working version
# 0.0.2: Sat 26 Oct 2013 16:47:54 EST
# Added uptime, timezone, install cluster and eeprom information
# 0.0.3: Sat 26 Oct 2013 17:22:47 EST
# Split reporting into separate reports
# 0.0.4: Mon 28 Oct 2013 19:21:46 EST
# Added code to check array of files in tar archive before extracting to speed up code
# 0.0.5: Wed  6 Nov 2013 19:21:58 EST
# Added initial security check code
# 0.0.6: Thu  7 Nov 2013 13:24:07 EST
# Added inetd service check
# 0.0.7: Fri  8 Nov 2013 08:49:05 EST
# Added Live Upgrade support
# 0.0.8: Fri  8 Nov 2013 13:14:20 EST
# Cleaned up output routine
# 0.0.9: Fri  8 Nov 2013 13:54:03 EST
# Added kernel module support
# 0.1.0: Fri  8 Nov 2013 22:44:50 EST
# Added package information
# 0.1.1: Fri  8 Nov 2013 23:23:06 EST
# Added patch information
# 0.1.2: Sat  9 Nov 2013 09:25:12 EST
# Added code to mask identifiable data
# 0.1.3: Sun 10 Nov 2013 09:24:01 EST
# Code clean up
# 0.1.4: Sun 22 Jun 2014 21:41:30 EST
# Added explorer listing and cleaned up code
# 0.1.5: Sun 22 Jun 2014 22:36:17 EST
# Added OBP function
# 0.1.6: Mon 23 Jun 2014 05:03:39 EST
# Added code to extract symlinks from explorer
# 0.1.7: Mon 23 Jun 2014 05:27:56 EST
# Fixed bugs with OBP version reporting
# 0.1.8: Thu 26 Jun 2014 08:28:25 EST
# Fixed option handling
# 0.1.9: Thu 26 Jun 2014 13:32:14 EST
# Replaced versionomy module as it doesn't handle long version strings
# 0.2.0: Thu 26 Jun 2014 14:27:45 EST
# Fixed Fcode determination for QLogic HBAs
# 0.2.1: Thu 26 Jun 2014 15:48:29 EST
# Split code out to make management easier
# 0.2.2: Thu 26 Jun 2014 16:40:33 EST
# Fixed tables for IO and CPU
# 0.2.3: Fri 27 Jun 2014 08:09:41 EST
# Fixed version comparing code
# 0.2.4: Sat 28 Jun 2014 09:41:23 EST
# Added code to check methods, information, and firmware directory exist
# 0.2.5: Sun 29 Jun 2014 15:32:38 EST
# Added support for determining XCP version on M10-4S
# 0.2.6: Sun 29 Jun 2014 18:06:58 EST
# Fixed bug with -s option
# 0.2.7: Mon 30 Jun 2014 11:39:07 EST
# Fixed available OBP determination
# 0.2.8: Mon 30 Jun 2014 16:19:43 EST
# Fixed QLogic firmware reporting
# 0.2.9: Tue  1 Jul 2014 12:56:54 EST
# Added initial T series and LDom support
# 0.3.0: Tue  1 Jul 2014 15:40:48 EST
# Improved LDom support
# 0.3.1: Wed  2 Jul 2014 11:44:20 EST
# Added masking for LDom configuration
# 0.3.2: Thu  3 Jul 2014 10:27:49 EST
# Added T2000 support and fixed bugs
# 0.3.3: Thu  3 Jul 2014 15:13:05 EST
# Added V120 support
# 0.3.4: Fri  4 Jul 2014 09:51:30 EST
# Added support for internal FC-AL on V series
# 0.3.5: Fri  4 Jul 2014 10:25:59 EST
# Fixed output for T5120
# 0.3.6: Fri  4 Jul 2014 12:23:43 EST
# Fixed output for M3000
# 0.3.7: Fri  4 Jul 2014 14:53:37 EST
# Cleaned up memory code
# 0.3.8: Fri  4 Jul 2014 17:23:29 EST
# General code cleanup
# 0.3.9: Sat  5 Jul 2014 06:27:38 EST
# Added FRU support
# 0.4.0: Sat  5 Jul 2014 15:51:31 EST
# Updated patch reporting
# 0.4.1: Mon  7 Jul 2014 09:18:04 EST
# Updated usage information
# 0.4.2: Wed  3 Sep 2014 16:19:08 AEST
# Fixed directory check
# 0.4.3: Wed  3 Sep 2014 16:54:08 AEST
# Use pigz to speed up decompression if available
# 0.4.4: Wed  3 Sep 2014 19:53:11 AEST
# Improved firmware URL reporting
# 0.4.5: Thu  4 Sep 2014 09:04:11 AEST
# Improved OBP version determination
# 0.4.6: Sun  7 Sep 2014 10:40:34 AEST
# Added initial support for multiple types of output
# 0.4.7: Sun  7 Sep 2014 18:43:30 AEST
# Cleaned up text file output
# 0.4.8: Mon  8 Sep 2014 04:00:18 AEST
# Initial PDF report support and various bug fixes
# 0.4.8: Mon  8 Sep 2014 15:12:10 AEST
# Cleaned up PDF and CPU output
# 0.4.9: Wed 10 Sep 2014 09:58:02 AEST
# Added Handbook processing and cleaned up Zone output
# 0.5.0: Wed 10 Sep 2014 15:10:27 AEST
# Cleaned up Handbook output
# 0.5.1: Wed 10 Sep 2014 21:09:31 AEST
# More Handbook handling improvements
# 0.5.2: Thu 11 Sep 2014 14:13:31 AEST
# Added support for parsing Facter output
# 0.5.3: Thu 11 Sep 2014 14:58:35 AEST
# Migrated most of masking code into output routine and fixed bugs with HBA handling
# 0.5.4: Fri 12 Sep 2014 09:29:38 AEST
# Updated Handbook parsing code
# 0.5.5: Fri 12 Sep 2014 15:02:00 AEST
# Added support to download Handbook files
# 0.5.6: Sun 14 Sep 2014 23:42:35 AEST
# Added initial Ansible and Puppet Fact reporting support
# 0.5.7: Mon 15 Sep 2014 21:20:14 AEST
# Added dmidecode support
# 0.5.8: Mon 15 Sep 2014 22:26:23 AEST
# Added support for M10-4S IO and CPU and cleaned up CPU output
# 0.5.9: Tue 16 Sep 2014 09:00:31 AEST
# Fixed package list for Solaris 11
# 0.6.0: Tue 16 Sep 2014 09:14:36 AEST
# Added package history for Solaris 11
# 0.6.1: Tue 16 Sep 2014 10:43:57 AEST
# Added package mediator information for Solaris 11
# 0.6.2: Tue 16 Sep 2014 13:14:59 AEST
# Added package properties and publisher information for Solaris 11
# 0.6.3: Tue 16 Sep 2014 15:51:11 AEST
# Added crypto information for Solaris 11
# 0.6.4: Tue 16 Sep 2014 16:09:50 AEST
# Added cups information
# 0.6.5: Tue 16 Sep 2014 16:57:00 AEST
# Added diskinfo function
# 0.6.6: Tue 16 Sep 2014 17:16:26 AEST
# Added swap information
# 0.6.7: Tue 16 Sep 2014 19:44:21 AEST
# Added mount information
# 0.6.8: Tue 16 Sep 2014 22:41:45 AEST
# Added code to reduce image size
# 0.6.9: Sun 30 Aug 2015 08:47:10 AEST
# Minor fix
# 0.7.0: Sun 30 Aug 2015 13:32:52 AEST
# Added check for parallel gzip
# 0.7.1: Mon 31 Aug 2015 21:37:09 AEST
# Improved code for putting images in PDF
# 0.7.2: Tue  1 Sep 2015 19:40:37 AEST
# Added support for M[5,6,7]-32
# 0.7.3: Thu  3 Sep 2015 15:02:22 AEST
# Fixed alignment for TOC
# 0.7.4: Thu  3 Sep 2015 21:32:26 AEST
# Fixed issues with PDF creation
# 0.7.5: Thu  3 Sep 2015 22:31:50 AEST
# Improved masking support and tables for PDF output
# 0.7.6: Fri  4 Sep 2015 08:07:13 AEST
# Improved image scaling
# 0.7.7: Fri  4 Sep 2015 10:14:14 AEST
# Added Solaris 11 BE support
# 0.7.8: Fri  4 Sep 2015 11:49:12 AEST

parsec.rb
#!/usr/bin/env ruby

# Name:         parsec (Explorer Parser)
# Version:      2.2.9
# Release:      1
# License:      CC-BA (Creative Commons By Attribution)
#               http://creativecommons.org/licenses/by/4.0/legalcode
# Group:        System
# Source:       N/A
# URL:          http://lateralblast.com.au/
# Distribution: UNIX
# Vendor:       Lateral Blast
# Packager:     Richard Spindler <richard@lateralblast.com.au>
# Description:  Ruby script for processing explorer

# Load gems

require 'rubygems'
require 'pathname'
require 'etc'
require 'date'

def install_gem(gem_name)
  puts "Information:\tInstalling #{gem_name}"
  %x[gem install #{gem_name}]
  Gem.clear_paths
  return
end

def install_pkg(pkg_name)
  $os_name = %[uname -a].chomp
  if $os_name.match(/SunOS/) and $os_name.match(/5\.11/)
    puts "Information:\tInstalling #{pkg_name}"
    %x[pkg install #{pkg_name}]
  else
    if $os_name.match(/Darwin/)
      if File.exist?("/usr/local/bin/brew")
        puts "Information:\tInstalling #{pkg_name}"
        %x[brew install #{pkg_name}]
      end
    end
  end
  return
end

begin
  require 'iconv'
rescue LoadError
  install_gem("iconv")
end
begin
  require 'getopt/long'
rescue LoadError
  install_gem("getopt")
end
begin
  require 'fileutils'
rescue LoadError
  install_gem("fileutils")
end
begin
  require 'hex_string'
rescue LoadError
  install_gem("hex_string")
end
begin
  require 'terminal-table'
rescue LoadError
  install_gem("terminal-table")
end
begin
  require 'unpack'
rescue LoadError
  install_gem("unpack")
end
begin
  require 'enumerator'
rescue LoadError
  install_gem("enumerator")
end
begin
  require 'prawn'
rescue LoadError
  install_gem("prawn")
end
begin
  require 'prawn/table'
rescue LoadError
  install_gem("prawn-table")
end
begin
  require 'fastimage'
rescue LoadError
  install_gem("fastimage")
end
begin
  require 'nokogiri'
rescue LoadError
  install_gem("nokogiri -- --use-system-libraries")
end
begin
  require 'rmagick'
  include Magick
rescue LoadError
  install_pkg("imagemagick")
  install_gem("rmagick")
  require 'rmagick'
  include Magick
end

# Script name

$script = $0

# Valid output formats

$valid_output_formats = [ 'table', 'pdf', 'pipe', 'csv', 'text' ]

# Handle CTRL-C more gracefully

Signal.trap("SIGINT") do
  exit 130
end

# Defaults for PDF

$default_font_size = 12
$section_font_size = 28
$heading_font_size = 18
$table_font_size   = 10

$company_name = "Lateral Blast Pty Ltd"
$author_name  = "Richard Spindler"

# Defaults for output

$output_format  = "pipe"

# Print usage

def print_help()
  switches     = []
  long_switch  = ""
  short_switch = ""
  help_info    = ""
  puts ""
  puts "Usage: "+$script
  puts ""
  file_array  = IO.readlines $0
  option_list = file_array.grep(/\[ "--/)
  option_list.each do |line|
    if !line.match(/file_array/)
      help_info    = line.split(/# /)[1]
      switches     = line.split(/,/)
      long_switch  = switches[0].gsub(/\[/,"").gsub(/\s+/,"")
      short_switch = switches[1].gsub(/\s+/,"")
      if long_switch.gsub(/\s+/,"").length < 7
        puts long_switch+",\t\t"+short_switch+"\t"+help_info
      else
        puts long_switch+",\t"+short_switch+"\t"+help_info
      end
    end
  end
  puts
  return
end

# Get command line arguments
# Print help if given none

if !ARGV[0]
  print_help()
  exit
end

# Try to make sure we have valid long switches

ARGV[0..-1].each do |switch|
  if switch.match(/^-[a-z]=/)
    puts "Invalid command line option: "+switch
    exit
  end
end

# Load methods

if Dir.exist?("./methods")
  file_list = Dir.entries("./methods")
  for file in file_list
    if file =~ /rb$/
      require "./methods/#{file}"
    end
  end
end

# Report types

report             = {}
report["aggr"]     = "Report on aggregate information"
report["all"]      = "Report on everything"
report["coreadm"]  = "Report on coreadm information"
report["cpu"]      = "Report on CPU information"
report["cron"]     = "Report on cron information"
report["crypto"]   = "Report on crypto information"
report["cups"]     = "Report on CUPS information"
report["disk"]     = "Report on disks"
report["dumpadm"]  = "Report on dumpadm information"
report["eeprom"]   = "Report on EEPROM"
report["elfsign"]  = "Report on elfsign information"
report["explorer"] = "Report on Explore information"
report["firmware"] = "Report on Firmware information"
report["fru"]      = "Report on FRU information"
report["fs"]       = "Report on filesystem information"
report["host"]     = "Report on host information"
report["inetadm"]  = "Report on inetadm information"
report["inetd"]    = "Report on inetd information"
report["inetinit"] = "Report on inetinit information"
report["io"]       = "Report on all IO"
report["ipmi"]     = "Report on IPMI information"
report["ldom"]     = "Report on LDom information"
report["link"]     = "Report on link information"
report["locale"]   = "Report on locale information"
report["login"]    = "Report on locale information"
report["lu"]       = "Report on Live Upgrade information"
report["kernel"]   = "Report on kernel information"
report["keyserv"]  = "Report on keyserv information"
report["memory"]   = "Report on Memory information"
report["modinfo"]  = "Report on Kernel Module information"
report["module"]   = "Report on Kernel Module information"
report["mount"]    = "Report on mount information"
report["ndd"]      = "Report on ndd information"
report["network"]  = "Report on network information"
report["ntp"]      = "Report on NTP information"
report["obp"]      = "Report on OBP information"
report["os"]       = "Report on OS information"
report["package"]  = "Report on package information"
report["pam"]      = "Report on PAM information"
report["password"] = "Report on password information"
report["patch"]    = "Report on patch information"
report["power"]    = "Report on power information"
report["security"] = "Report on Security information"
report["sendmail"] = "Report on Sendmail settings"
report["sensors"]  = "Report on Sensor information"
report["serial"]   = "Report on Chassis Serial information"
report["serials"]  = "Report on Component Serial information"
report["services"] = "Report on Services information"
report["slots"]    = "Report on Upgradeable slot information"
report["snmp"]     = "Report on SNMP information"
report["ssh"]      = "Report on SSH information"
report["su"]       = "Report on SU information"
report["suspend"]  = "Report on Suspend information"
report["syslog"]   = "Report on syslog information"
report["system"]   = "Report on system information"
report["swap"]     = "Report on swap information"
report["svm"]      = "Report on Solaris Volume Manager information"
report["tcp"]      = "Report on TCP information"
report["telnet"]   = "Report on Telnet information"
report["udp"]      = "Report on UDP information"
report["vnic"]     = "Report on VNIC information"
report["veritas"]  = "Report on Veritas information"
report["zfs"]      = "Report on ZFS information"
report["zone"]     = "Report on Zone information"
#report[""]=""

# Options

# Process options

begin
  option = Getopt::Long.getopts(
    [ "--prefix",       "-b", Getopt::REQUIRED ],   # Set base/prefix directory
    [ "--customer",     "-C", Getopt::REQUIRED ],   # Set customer name
    [ "--date",         "-T", Getopt::REQUIRED ],   # Set date (used in conjunction with list)
    [ "--dir",          "-d", Getopt::BOOLEAN ],    # Specify directory where explorers are (default is ./explorers)
    [ "--dodisks",      "-D", Getopt::BOOLEAN ],    # Do disks (reports information on all disks)
    [ "--format",       "-f", Getopt::REQUIRED ],   # Output format
    [ "--list",         "-l", Getopt::BOOLEAN ],    # List explorers (also lists dmidecodes and facters if present)
    [ "--masked",       "-m", Getopt::BOOLEAN ],    # Mask hostnames, IPs, MAC addresses etc
    [ "--model",        "-M", Getopt::REQUIRED ],   # Set model (used in conjunction with list)
    [ "--input",        "-i", Getopt::REQUIRED ],   # Input file
    [ "--output",       "-o", Getopt::REQUIRED ],   # Output file
    [ "--pause",        "-p", Getopt::BOOLEAN ],    # Pause between each report when running against all hosts (useful for debugging)
    [ "--report",       "-R", Getopt::REQUIRED ],   # Report type (e.g. all, cpu, memory)
    [ "--server",       "-s", Getopt::REQUIRED ],   # Server to run explorer report for
    [ "--type",         "-t", Getopt::REQUIRED ],   # Type of report (default is explorer)
    [ "--date",         "-T", Getopt::REQUIRED ],   # Set date (used in conjunction with list)
    [ "--temp",         "-w", Getopt::REQUIRED ],   # Work directory
    [ "--usage",        "-u", Getopt::REQUIRED ],   # Display usage information
    [ "--use",          "-U", Getopt::REQUIRED ],   # Override defaults, e.g. use gzip rather than pgiz
    [ "--year",         "-Y", Getopt::REQUIRED ],   # Set year(used in conjunction with list)
    [ "--help",         "-h", Getopt::BOOLEAN ],    # Display help information
    [ "--verbose",      "-v", Getopt::BOOLEAN ],    # Verbose output
    [ "--version",      "-V", Getopt::BOOLEAN ],    # Display version
    [ "--changelog",    "-c", Getopt::BOOLEAN ]     # Print changelog
 )
rescue
  print_help()
  exit
end

# Print Changelog

def print_changelog()
  if File.exist?("changelog")
    changelog = File.readlines("changelog")
    changelog = changelog.reverse
    changelog.each_with_index do |line, index|
      line = line.gsub(/^# /,"")
      if line.match(/^[0-9]/)
        puts line
        puts changelog[index-1].gsub(/^# /,"")
        puts
      end
    end
  end
  return
end

def check_output_format()
  if !$valid_output_formats.grep(/#{$output_format}/)
    puts "Invalid output format"
    exit
  end
end

opt = option

# Print examples

def print_examples(usage)
  match = 0
  if usage.match(/example|report|server|cpu/)
    match = 1
    puts "Run a CPU report against a specific host"
    puts
    puts "$ #{$script} --server=hostname --report=cpu"
    puts
    puts "Run a CPU report against all hosts"
    puts
    puts "$ #{$script} --server=all --report=cpu"
    puts
  end
  if usage.match(/example|report|server|cpu/)
    match = 1
    puts "Run a all reports against a specific host"
    puts
    puts "$ #{$script} --server=hostname --report=all"
    puts
    puts "Run all reports against all hosts"
    puts
    puts "$ #{$script} --server=all --report=all"
    puts
  end
  if usage.match(/example|report|server|cpu|format/)
    match = 1
    puts "Run all reports against a specific host and output a PDF report"
    puts "(will create a file based on the hostname and report in the outputs directory)"
    puts
    puts "$ #{$script} --server=hostname --report=all --format=pdf"
    puts
    puts "Run all reports against a specific host and output PDF a report to a specific file"
    puts
    puts "$ #{$script} --server=hostname --report=all --format=pdf --output=hostename.pdf"
    puts
  end
  if usage.match(/example|report|server|cpu|format/)
    match = 1
    puts "Run a all reports against a specific host and output it in a format that can be piped into another command"
    puts
    puts "$ #{$script} --server=hostname --report=all --format=pipe"
    puts
  end
  if usage.match(/example|report|server|cpu|format/)
    match = 1
    puts "Run a all reports against a specific host and output it in a table format"
    puts
    puts "$ #{$script} --server=hostname --report=all --format=table"
    puts
  end
  if match == 0
    print_help()
  end
  return
end

# Print usage

def print_usage(usage,report)
  case usage
  when /reports/
    puts
    report.each do |type,info|
      if type.length < 7
        puts type+":\t\t"+info
      else
        puts type+":\t"+info
      end
    end
  else
    puts
    print_examples(usage)
  end
  puts
  return
end

# Overide defaults

if option["use"]
  use_flag = option["use"]
  case use_flag
  when /star/
    $tar_bin = %x[which star].chomp
    if !$tar_bin.match(/star/) or $tar_bin.match(/no star/)
      $tar_bin = %x[which tar].chomp
    end
  when /tar/
    $tar_bin = %x[which tar].chomp
  when /pigz/
    $gzip_bin = %x[which pigz].chomp
    if !$gzip_bin.match(/pigz/) or $gzip_bin.match(/no pigz/)
      $gzip_bin = %x[which gzip].chomp
    end
  when /gzip/
    $gzip_bin = %x[which gzip].chomp
  end
end

# Print help

if option["help"]
  print_help()
  exit
end

if option["usage"]
  usage = option["usage"]
  print_usage(usage,report)
  exit
end

# Enable verbose mode

if option["verbose"]
  $verbose_mode = 1
  puts "Operating in verbose mode"
end

# get server name

if option["server"]
  host_name = option["server"].downcase
  if !option["report"]
    option["report"] = "all"
    if $verbose_mode == 1
      puts "reporting on all elements of "+host_name
    end
  end
end

# Get model type

if option["model"]
  search_model = option["model"]
else
  search_model = ""
end

# Get date string

if option["date"]
  search_date = option["date"]
  if !search_date.match(/latest|last|earliest|first/)
    begin
      search_date = Date.parse(search_date).to_s
    rescue
      puts "Invalid date"
      exit
    end
  end
else
  if option["list"]
    search_date = ""
  else
    search_date = "latest"
  end
end

# Get year string

if option["year"]
  search_year = option["year"]
  if search_year.length == 2
    search_year = "20"+search_year
  end
  if search_year.match(/[a-z,A-Z]/)
    puts "Invalid year"
    exit
  end
else
  search_year = ""
end

# Check local config

if !option["help"] and !option["version"]
  check_local_config()
end

if option["prefix"]
  $base_dir = option["prefix"]
else
  $base_dir = Dir.pwd
end

# Print version

if option["version"]
  print_version()
  exit
end

# Print changelog

if option["changelog"]
  print_changelog()
  exit
end

# Set format of report

if option["format"]
  $output_format = option["format"].downcase
  $output_format = $output_format.gsub(/text/,"pipe")
  check_output_format()
else
  $output_format = "pipe"
  if $verbose_mode == 1
    puts "Setting output type to "+$output_format
  end
end

# Pause mode

if option["pause"]
  pause_mode = 1
else
  pause_mode = 0
end

# Mask data

if option["masked"]
  $masked = 1
  if $verbose_mode == 1
    puts "Masking output"
  end
end

# Set explorer directory

if option["dir"]
  $exp_dir = opt["dir"]
else
  if !$exp_dir.match(/[A-z]/)
    $exp_dir = Dir.pwd+"/explorers"
  end
end

# Set type

if option["type"]
  input_type = option["type"]
else
  input_type = "explorer"
end

# Set output file

if option["output"]
  $output_file = option["output"]
  $output_dir  = File.dirname($output_file)
  if !File.directory?($output_dir) and !File.symlink?($output_dir)
    Dir.mkdir($output_dir)
  end
  if $verbose_mode == 1 and !host_name.match(/^all$/)
    puts "Setting output file to: "+$output_file
  end
  if File.exist?($output_file)
    File.delete($output_file)
    FileUtils.touch($output_file)
  else
    FileUtils.touch($output_file)
  end
else
  $output_file = ""
  $output_dir  = $base_dir+"/output"
  if !File.directory?($output_dir) and !File.symlink?($output_dir)
    Dir.mkdir($output_dir)
  end
  if $output_format.match(/pdf/)
    if !host_name.match(/^all$/)
      $output_file = $output_dir+"/"+host_name+".txt"
    end
  end
end

# List explorers

if option["list"]
  if input_type.match(/all|explorer/)
    if !host_name
      host_name = ""
    end
    list_explorers(search_model,search_date,search_year,host_name)
  end
  if input_type.match(/all|facter/)
    list_facters()
  end
  if input_type.match(/all|dmidecode/)
    list_dmidecodes()
  end
  exit
end

# Report on all disk information

if option["dodisks"]
  $do_disks = 1
  if $verbose_mode == 1
    puts "Enabling full disk reporting"
  end
end

# Handle filename and hostname options

if option["input"]
  $exp_file = option["input"]
  if !File.exist?($exp_file) and !File.symlink?($exp_file)
    puts
    puts "Input file: "+$exp_file+" does not exist"
    exit
  end
  host_name  = get_hostname_from_explorer_File()
  exp_data[host_name]["file"] = $exp_file
  host_names = []
  host_names.push(host_name)
else
  if !option["server"]
    if !option["help"]
      puts
      puts "Input file or hostname not specified"
    end
    print_help()
    exit
  end
  $exp_dir    = $base_dir.chomp()
  $exp_dir    = $exp_dir+"/explorers"
  search_name = option["server"]
  file_list   = get_explorer_file_list(search_model,search_date,search_year,search_name)
end

# Set work directory

if option["temp"]
  $work_dir = option["temp"]
else
  if !$work_dir.match(/[A-z]/)
    $work_dir = "/tmp"
  end
end

# Get report type

if option["report"]
  $report_type = option["report"]
  if !option["input"] and !option["server"]
    puts "Input file or home name not specified"
    print_help()
    exit
  end
end

# Get customer name

if option["customer"]
  if $masked == "1"
    customer_name = "Masked"
  else
    customer_name = option["customer"]
  end
else
  customer_name = ""
end

# Handle explorer output

if input_type.match(/explorer/)
 handle_explorer(report,file_list,search_model,search_date,search_year,search_name)
end
